### ISSUE TITLE
Package not installing to correct location

### ISSUE THREAD

**Issue by balbuf** [oomphinc/composer-installers-extender#16]
_Originally posted by @Nick-Hope in https://github.com/oomphinc/composer-installers-extender/issues/6#issuecomment-437649730_

I am still suffering this problem with Composer 1.7.3 (in Ubuntu 18.04). The `harvesthq/chosen` library will not install in my `web/libraries` directory, no matter what I try; it always goes to the `vendor` directory. It won't even go there if I run `composer update` or `composer install` twice. As far as I can tell I have tried everything suggested in this thread. "Chosen" was specifically mentioned in [this comment above by landsman](https://github.com/oomphinc/composer-installers-extender/issues/6#issuecomment-295750819), so maybe it is something specific to that package. Here's my composer.json in full. Note that I have already tried without `harvesthq/chosen` specified in the `installer-paths` section.

```
{
    "type": "project",
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        },
        {
            "type": "composer",
            "url": "https://asset-packagist.org"
        },
        {
            "type": "package",
            "package": {
                "name": "jackmoore/colorbox",
                "version": "1.6.4",
                "type": "drupal-library",
                "dist": {
                    "url": "https://github.com/jackmoore/colorbox/archive/1.6.4.zip",
                    "type": "zip"
                }
            }
        }
    ],
    "require": {
        "composer/installers": "^1.5",
        "oomphinc/composer-installers-extender": "^1.1.2",
        "cweagans/composer-patches": "^1.6.5",
        "drupal-composer/drupal-scaffold": "^2.5",
        "drupal/admin_toolbar": "^1.24",
        "drupal/allowed_formats": "^1.1",
        "drupal/better_exposed_filters": "^3.0@alpha",
        "drupal/chosen": "^2.6",
        "drupal/colorbox": "^1.4",
        "drupal/console": "^1.0.2",
        "drupal/core": "^8.6.3",
        "drupal/ctools": "^3.0",
        "drupal/facets": "1.1",
        "drupal/field_permissions": "^1.0@rc",
        "drupal/geolocation": "2.x-dev",
        "drupal/hierarchical_term_formatter": "^1.0@beta",
        "drupal/image_style_quality": "^1.3",
        "drupal/media_taxonomy_filter": "2.x-dev",
        "drupal/pathauto": "^1.3",
        "drupal/quicktabs": "^3.0@alpha",
        "drupal/search_api": "^1.10",
        "drupal/search_api_autocomplete": "^1.1",
        "drupal/shs": "^1.0@alpha",
        "drupal/superfish": "^1.2",
        "drupal/taxonomy_manager": "^1.0@alpha",
        "drupal/taxonomy_menu": "^3.3",
        "drupal/token": "^1.5",
        "drupal/tvi": "^1.0@beta",
        "drupal/video_embed_field": "^2.0@alpha",
        "drupal/views_arg_order_sort": "1.x-dev",
        "drupal/views_contextual_filters_or": "^1.0",
        "drupal/views_infinite_scroll": "^1.5",
        "bower-asset/jquery-ui-slider-pips": "^1.11.4",
        "jackmoore/colorbox": "^1.6.4",
        "vlucas/phpdotenv": "^2.5.1",
        "webflo/drupal-finder": "^1.1",
        "webmozart/path-util": "^2.3",
        "wikimedia/composer-merge-plugin": "^1.4.1",
        "harvesthq/chosen": "^1.8"
    },
    "require-dev": {
        "drupal/devel": "^1.2",
        "webflo/drupal-core-require-dev": "^8.6.2"
    },
    "conflict": {
        "drupal/drupal": "*"
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "config": {
        "preferred-install": "dist",
        "autoloader-suffix": "Drupal8"
    },
    "autoload": {
        "psr-4": {
            "Drupal\\Core\\Composer\\": "core/lib/Drupal/Core/Composer"
        }
    },
    "scripts": {
        "pre-autoload-dump": "Drupal\\Core\\Composer\\Composer::preAutoloadDump",
        "post-autoload-dump": "Drupal\\Core\\Composer\\Composer::ensureHtaccess",
        "post-package-install": "Drupal\\Core\\Composer\\Composer::vendorTestCodeCleanup",
        "post-package-update": "Drupal\\Core\\Composer\\Composer::vendorTestCodeCleanup",
        "drupal-phpunit-upgrade-check": "Drupal\\Core\\Composer\\Composer::upgradePHPUnit",
        "drupal-phpunit-upgrade": "@composer update phpunit/phpunit phpspec/prophecy symfony/yaml --with-dependencies --no-progress",
        "phpcs": "phpcs --standard=core/phpcs.xml.dist --runtime-set installed_paths $($COMPOSER_BINARY config vendor-dir)/drupal/coder/coder_sniffer --",
        "phpcbf": "phpcbf --standard=core/phpcs.xml.dist --runtime-set installed_paths $($COMPOSER_BINARY config vendor-dir)/drupal/coder/coder_sniffer --"
    },
    "extra": {
        "_readme": [
            "By default Drupal loads the autoloader from ./vendor/autoload.php.",
            "To change the autoloader you can edit ./autoload.php.",
            "This file specifies the packages.drupal.org repository.",
            "You can read more about this composer repository at:",
            "https://www.drupal.org/node/2718229"
        ],
        "enable-patching": true,
        "patchLevel": {
            "drupal/core": "-p2"
        },
        "installer-types": [
            "bower-asset",
            "npm-asset"
        ],
        "installer-paths": {
            "web/core": ["type:drupal-core"],
            "web/libraries/{$name}": [
                "type:drupal-library",
                "type:bower-asset",
                "type:npm-asset",
                "harvesthq/chosen"
            ],
            "web/modules/contrib/{$name}": ["type:drupal-module"],
            "web/profiles/contrib/{$name}": ["type:drupal-profile"],
            "web/themes/contrib/{$name}": ["type:drupal-theme"],
            "drush/contrib/{$name}": ["type:drupal-drush"],
            "web/modules/custom/{$name}": ["type:drupal-custom-module"],
            "web/themes/custom/{$name}": ["type:drupal-custom-theme"]
        },
        "drupal-scaffold": {
            "initial": {
                ".editorconfig": "../.editorconfig",
                ".gitattributes": "../.gitattributes"
            }
        },
        "merge-plugin": {
            "include": [
                "core/composer.json",
                "web/modules/custom/*/composer.json"
            ],
            "recurse": true,
            "replace": false,
            "merge-extra": false,
            "ignore-duplicates": true
        },
        "patches": {
            "drupal/core": {
                "Render action button above parent entity": "https://www.drupal.org/files/issues/preview-comment-aoutput-below-2927874-2.diff",
                "Hide undefined index errors on all pages": "https://www.drupal.org/files/issues/filterpluginbase-filter-notices.patch"
            },
            "drupal/geolocation": {
                "Respect initial zoom value": "https://www.drupal.org/files/issues/2018-11-02/geolocation-respect-initial-zoom-level-2945234-60.patch"
            },
            "drupal/facets": {
                "Display titles on slider facets": "https://www.drupal.org/files/issues/2018-10-18/3002580_6.patch"
            },
            "drupal/views_arg_order_sort": {
                "Fix SQL injection attack vulnerability": "https://www.drupal.org/files/issues/2018-10-19/views-arg-order-sort-sql-injection-fix-2982591-5.patch"
            }
        }
    }
}
```

**Comment by balbuf** [comment]
@Nick-Hope the `harvesthq/chosen` package [has a type of `library`](https://github.com/harvesthq/chosen-package/blob/master/composer.json#L30), so you need to add that to the `"installer-types"` array. The installer extender will only act on types defined in this array.

You might also look at my other package https://github.com/balbuf/drupal-libraries-installer which makes installing arbitrary front-end libraries much easier for this purpose, but it lacks the ability to use version constraints / resolution.

Let me know if you have any further issues!

**Comment by Nick-Hope** [comment]
Thank you for the reply and apologies for mistakenly using that other thread.

In my `composer.json` file I added `"library"` to the installer types and removed the `"package"` section for `harvest/chosen`. I wasn't sure from your comment whether this alone would work or if I would also need your `drupal-libraries-installer` package. It didn't work (`chosen` still went into `vendor`), so for the time being I've decided to quit while I'm winning. Here's my full `composer.json` file that puts the `chosen` library where I want it in `web\libraries`:
```
{
    "type": "project",
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        },
        {
            "type": "composer",
            "url": "https://asset-packagist.org"
        },
        {
            "type": "package",
            "package": {
                "name": "jackmoore/colorbox",
                "version": "1.6.4",
                "type": "drupal-library",
                "dist": {
                    "url": "https://github.com/jackmoore/colorbox/archive/1.6.4.zip",
                    "type": "zip"
                }
            }
        },
        {
            "type": "package",
            "package": {
                "name": "harvesthq/chosen",
                "version": "1.8.7",
                "type": "drupal-library",
                "dist": {
                    "url": "https://github.com/harvesthq/chosen/releases/download/v1.8.7/chosen_v1.8.7.zip",
                    "type": "zip"
                }
            }
        }
    ],
    "require": {
        "composer/installers": "^1.5",
        "oomphinc/composer-installers-extender": "^1.1.2",
        "cweagans/composer-patches": "^1.6.5",
        "drupal-composer/drupal-scaffold": "^2.5",
        "drupal/admin_toolbar": "^1.24",
        "drupal/allowed_formats": "^1.1",
        "drupal/better_exposed_filters": "^3.0@alpha",
        "drupal/chosen": "^2.6",
        "drupal/colorbox": "^1.4",
        "drupal/console": "^1.0.2",
        "drupal/core": "^8.6.3",
        "drupal/ctools": "^3.0",
        "drupal/facets": "1.1",
        "drupal/field_permissions": "^1.0@rc",
        "drupal/geolocation": "2.x-dev",
        "drupal/hierarchical_term_formatter": "^1.0@beta",
        "drupal/image_style_quality": "^1.3",
        "drupal/media_taxonomy_filter": "2.x-dev",
        "drupal/pathauto": "^1.3",
        "drupal/quicktabs": "^3.0@alpha",
        "drupal/search_api": "^1.10",
        "drupal/search_api_autocomplete": "^1.1",
        "drupal/shs": "^1.0@alpha",
        "drupal/superfish": "^1.2",
        "drupal/taxonomy_manager": "^1.0@alpha",
        "drupal/taxonomy_menu": "^3.3",
        "drupal/token": "^1.5",
        "drupal/tvi": "^1.0@beta",
        "drupal/video_embed_field": "^2.0@alpha",
        "drupal/views_arg_order_sort": "1.x-dev",
        "drupal/views_contextual_filters_or": "^1.0",
        "drupal/views_infinite_scroll": "^1.5",
        "bower-asset/jquery-ui-slider-pips": "^1.11.4",
        "jackmoore/colorbox": "^1.6.4",
        "vlucas/phpdotenv": "^2.5.1",
        "webflo/drupal-finder": "^1.1",
        "webmozart/path-util": "^2.3",
        "wikimedia/composer-merge-plugin": "^1.4.1",
        "harvesthq/chosen": "^1.8.7"
    },
    "require-dev": {
        "drupal/devel": "^1.2",
        "webflo/drupal-core-require-dev": "^8.6.2"
    },
    "conflict": {
        "drupal/drupal": "*"
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "config": {
        "preferred-install": "dist",
        "autoloader-suffix": "Drupal8"
    },
    "autoload": {
        "psr-4": {
            "Drupal\\Core\\Composer\\": "core/lib/Drupal/Core/Composer"
        }
    },
    "scripts": {
        "pre-autoload-dump": "Drupal\\Core\\Composer\\Composer::preAutoloadDump",
        "post-autoload-dump": "Drupal\\Core\\Composer\\Composer::ensureHtaccess",
        "post-package-install": "Drupal\\Core\\Composer\\Composer::vendorTestCodeCleanup",
        "post-package-update": "Drupal\\Core\\Composer\\Composer::vendorTestCodeCleanup",
        "drupal-phpunit-upgrade-check": "Drupal\\Core\\Composer\\Composer::upgradePHPUnit",
        "drupal-phpunit-upgrade": "@composer update phpunit/phpunit phpspec/prophecy symfony/yaml --with-dependencies --no-progress",
        "phpcs": "phpcs --standard=core/phpcs.xml.dist --runtime-set installed_paths $($COMPOSER_BINARY config vendor-dir)/drupal/coder/coder_sniffer --",
        "phpcbf": "phpcbf --standard=core/phpcs.xml.dist --runtime-set installed_paths $($COMPOSER_BINARY config vendor-dir)/drupal/coder/coder_sniffer --"
    },
    "extra": {
        "_readme": [
            "By default Drupal loads the autoloader from ./vendor/autoload.php.",
            "To change the autoloader you can edit ./autoload.php.",
            "This file specifies the packages.drupal.org repository.",
            "You can read more about this composer repository at:",
            "https://www.drupal.org/node/2718229"
        ],
        "enable-patching": true,
        "patchLevel": {
            "drupal/core": "-p2"
        },
        "installer-types": [
            "bower-asset",
            "npm-asset"
        ],
        "installer-paths": {
            "web/core": ["type:drupal-core"],
            "web/libraries/{$name}": [
                "type:drupal-library",
                "type:bower-asset",
                "type:npm-asset"
            ],
            "web/modules/contrib/{$name}": ["type:drupal-module"],
            "web/profiles/contrib/{$name}": ["type:drupal-profile"],
            "web/themes/contrib/{$name}": ["type:drupal-theme"],
            "drush/contrib/{$name}": ["type:drupal-drush"],
            "web/modules/custom/{$name}": ["type:drupal-custom-module"],
            "web/themes/custom/{$name}": ["type:drupal-custom-theme"]
        },
        "drupal-scaffold": {
            "initial": {
                ".editorconfig": "../.editorconfig",
                ".gitattributes": "../.gitattributes"
            }
        },
        "merge-plugin": {
            "include": [
                "core/composer.json",
                "web/modules/custom/*/composer.json"
            ],
            "recurse": true,
            "replace": false,
            "merge-extra": false,
            "ignore-duplicates": true
        },
        "patches": {
            "drupal/core": {
                "Render action button above parent entity": "https://www.drupal.org/files/issues/preview-comment-aoutput-below-2927874-2.diff",
                "Hide undefined index errors on all pages": "https://www.drupal.org/files/issues/filterpluginbase-filter-notices.patch"
            },
            "drupal/geolocation": {
                "Respect initial zoom value": "https://www.drupal.org/files/issues/2018-11-02/geolocation-respect-initial-zoom-level-2945234-60.patch"
            },
            "drupal/facets": {
                "Display titles on slider facets": "https://www.drupal.org/files/issues/2018-10-18/3002580_6.patch"
            },
            "drupal/views_arg_order_sort": {
                "Fix SQL injection attack vulnerability": "https://www.drupal.org/files/issues/2018-10-19/views-arg-order-sort-sql-injection-fix-2982591-5.patch"
            }
        }
    }
}```

**Comment by balbuf** [comment]
Hey @Nick-Hope, this works because you've explicitly declared the `harvesthq/chosen` package with a type of `drupal-library` (in your `repositories` config). Your original `composer.json` you shared should work if you just add `library` to the `installer-types` array.

The way you have it now you shouldn't even need this package (`oomphinc/composer-installers-extender`), because `composer/installers` supports the `drupal-library` package type OOTB. You'd only need this package if you went with your original `composer.json` and needed to specify a custom install path for the `library` package type.

If you were to use `balbuf/drupal-libraries-installer`, you also wouldn't need this package, but you would still need `composer/installers` and you'd need to specify your libraries in a separate place in your `composer.json` file (as described in the readme).

Anyway, it sounds like you've got this figured out, so I will go ahead and close this issue.
